src/models/StyledComponent.js

```diff
@@ -29,224 +30,225 @@ type BaseState = {
-          nextProps,
-          prevState.theme,
+          this.props,
+          nextTheme,
           this.constructor.defaultProps
         )
+
         const generatedClassName = this.generateAndInjectStyles(
           theme,
-          nextProps
+          this.props
         )
 
-        return { theme, generatedClassName }
+        this.setState({ theme, generatedClassName })
       })
+    } else {
+      // eslint-disable-next-line react/prop-types
+      const theme = this.props.theme || EMPTY_OBJECT
+      const generatedClassName = this.generateAndInjectStyles(theme, this.props)
+      this.setState({ theme, generatedClassName })
     }
+  }
 
-    componentWillUnmount() {
-      this.unsubscribeFromContext()
+  componentWillReceiveProps(nextProps: { theme?: Theme, [key: string]: any }) {
+    // If this is a statically-styled component, we don't need to listen to
+    // props changes to update styles
+    const { componentStyle } = this.constructor
+    if (componentStyle.isStatic) {
+      return
     }
 
-    render() {
-      // eslint-disable-next-line react/prop-types
-      const { innerRef } = this.props
-      const { generatedClassName } = this.state
-      const { styledComponentId, target } = this.constructor
-
-      const isTargetTag = isTag(target)
-
-      const className = [
-        // eslint-disable-next-line react/prop-types
-        this.props.className,
-        styledComponentId,
-        this.attrs.className,
-        generatedClassName,
-      ]
-        .filter(Boolean)
-        .join(' ')
-
-      const baseProps: any = {
-        ...this.attrs,
-        className,
-      }
+    this.setState(prevState => {
+      const theme = determineTheme(
+        nextProps,
+        prevState.theme,
+        this.constructor.defaultProps
+      )
+      const generatedClassName = this.generateAndInjectStyles(theme, nextProps)
 
-      if (isStyledComponent(target)) {
-        baseProps.innerRef = innerRef
-      } else {
-        baseProps.ref = innerRef
-      }
+      return { theme, generatedClassName }
+    })
+  }
 
-      const propsForElement = Object.keys(this.props).reduce(
-        (acc, propName) => {
-          // Don't pass through non HTML tags through to HTML elements
-          // always omit innerRef
-          if (
-            propName !== 'innerRef' &&
-            propName !== 'className' &&
-            (!isTargetTag || validAttr(propName))
-          ) {
-            // eslint-disable-next-line no-param-reassign
-            acc[propName] = this.props[propName]
-          }
-
-          return acc
-        },
-        baseProps
-      )
+  componentWillUnmount() {
+    this.unsubscribeFromContext()
+  }
+
+  render() {
+    // eslint-disable-next-line react/prop-types
+    const { innerRef } = this.props
+    const { generatedClassName } = this.state
+    const { styledComponentId, target } = this.constructor
+
+    const isTargetTag = isTag(target)
+
+    const className = [
```

Comment:
We could replace this one with a simpler concatenation, especially since extra white spaces don’t matter we don’t have to do this too correctly